### PR TITLE
Feat: memo expected

### PR DESCRIPTION
**Pull request by MarvinJanssen**
This PR adds two new commands to the send-many-stx-cli tool.

1. `send-many-memo-safe`
The 'safe' counterpart of `send-many-memo` checks for the existence of a `memo-expected` contract for each recipient. If it exists, the transfer will be aborted if the corresponding memo is empty or missing.
2. `deploy-contract`
A utility command to simplify deploying contracts related to the STX bulk transfer tool. It deploys the contract on the address of the provided private key. Valid choices are: send-many, send-many-memo, and memo-expected. The memo-expected contract is an empty contract that is checked by the `send-many-memo-safe` command mentioned above.

**Rationale:**
- Some exchanges use this tool to send bulk transfers for withdrawals. Some of those withdrawals end up going to different exchanges. It is a relatively common occurrence that multiple exchange users request a withdrawal to a different exchange at the same time, whilst forgetting to input a memo. The exchanges combine withdrawals to the same principal + memo to save on fees. The problem is that the incoming exchange sees one large transfer and has a hard time figuring out what belongs to whom.
- After a lot of debate, this seems like the safest and least intrusive option. The exchanges that use this tool can switch to using the `-safe` counterpart and all will benefit.
- As the tool does not keep state, the result of `checkMemoExpected()` is not cached. Could be nice to do that at some point, although the large majority of the addresses will never deploy a "memo-required". Exchanges will probably want to use their local API for `nodeUrl`.

Let me know if there are any questions or we can hop on a call.

Still WIP, looking for input. I also assume we have to change the contract path for dist? We should probably also add JSON support for `send-many` and deduplicate code.

### CODE REVIEW

File: src/commands/send-many-memo-safe.ts
Comment (jcnelson) [inline src/commands/send-many-memo-safe.ts:176]: Maybe I've been staring at Rust code for too long, but shouldn't this be `memoExpectedRecipients.length > 0`?

File: src/commands/send-many-memo-safe.ts
Comment (jcnelson) [inline src/commands/send-many-memo-safe.ts:46]: Do we want to throw an exception here if `entry.response.status` is a 500 error?  Otherwise, we could have a false negative on our hands -- a recipient expected a memo, but this `.filter()` didn't report it.

**Review by jcnelson — APPROVED** [review]
This overall LGTM.  I like the approach -- makes it really easy to for anyone to tag prominent recipients (e.g. exchanges) as requiring a memo.

Just had a couple of minor questions.

File: src/commands/send-many-memo-safe.ts
Comment (MarvinJanssen) [reply src/commands/send-many-memo-safe.ts:176]: A non-zero integer is truthy in JavaScript 😁. We can make it explicit though.

File: src/commands/send-many-memo-safe.ts
Comment (MarvinJanssen) [reply src/commands/send-many-memo-safe.ts:46]: `fetch()` will throw in this case.

File: src/commands/send-many-memo-safe.ts
Comment (hstove) [inline src/commands/send-many-memo-safe.ts:38]: I think that this will compile to `false` if the memo is undefined, because `undefined === 0` will be false. So, if the memo is `undefined`, this will not get caught, right?

**Review by hstove — CHANGES_REQUESTED** [review]
I think this is a great addition, but I'm worried about the one comment I left leaving an edge case uncaught.

At a meta level, I'm no longer employed at Hiro, so I'm really not sure how we should move forward here. Once this is changed, I'm fine with approving, but it might not be appropriate for me to merge and publish to npm. cc @diwakergupta

File: src/commands/send-many-memo-safe.ts
Comment (MarvinJanssen) [reply src/commands/send-many-memo-safe.ts:38]: You're right, I wrongly assumed the memo would always be set since a split of `"a,b,"` becomes `["a","b",""]` but if the user leaves out the second comma then memo becomes `undefined`. Best to add the edge-case in the filter instead of defaulting to an empty string when splitting the recipients, I think.

**Comment by diwakergupta** [comment]
> I think this is a great addition, but I'm worried about the one comment I left leaving an edge case uncaught.
> 
> At a meta level, I'm no longer employed at Hiro, so I'm really not sure how we should move forward here. Once this is changed, I'm fine with approving, but it might not be appropriate for me to merge and publish to npm. cc @diwakergupta

Thanks Hank. I think we can decouple the development for this tool from the npm scope ownership & publication process.

In the meantime, if this PR has 2 approvals, let's merge it in while we figure out a durable ownership and process for this going forward.

**Comment by MarvinJanssen** [comment]
Any movement on this?

**Comment by diwakergupta** [comment]
Let's merge it once @zone117x approves, just so we have another set of eyes. Matt, once merged let's publish manually for now and then we can work with devops to automate publishing from this repo.

**Review by zone117x — APPROVED** [review]

**Comment by zone117x** [comment]
Published https://www.npmjs.com/package/@stacks/send-many-stx-cli/v/1.2.0

**Pull request merged**
